HF QIE10 for 2016

kpedro88 opened the pull request:
The long-awaited (and rather large) implementation of QIE10 digitization for HF.

Most of the changes are in the digitization and packer/unpacker code to support the QIE10 data frames.

For the 2016 scenario, the digitizer supports a "mixed" case where some HF cells are digitized with QIE8 and some are digitized with QIE10. This is configured by a combination of python flags (with a corresponding new `run2_HF_2016` sub-era) and database conditions (the new `QIEType` flag, see #12473, #12905, #13004). The QIE10 digi collection will be stored for private studies, but it will not be used for 2016 RECO.

This should have no impact on existing workflows. It will only be activated in combination with the Extended2016 geometry in #13139, related changes in #13205, and an upcoming PR from @bsunanda for some minor related simulation changes. There will also be database changes needed, which will be forthcoming (current tests have used customized hardcode conditions to simulate the necessary conditions).

Some QIE10 features are not available yet:
1. TDC simulation
2. PreMixDigis functionality
3. Some zero suppression-related functions
However, for 2016 we only need to keep the RAW data from the QIE10s, so these features can be implemented later (targeted for 2017).

Attention: @abdoulline, @bsunanda, @jmmans, @awhitbeck

Spudmeister committed:
Updated HcalUnpacker to handle QIE10 and 11

Spudmeister committed:
Added some convenience functions

Spudmeister committed:
Fixed rawToDigi to add QIE11 to event and killed off verbosity

Spudmeister committed:
Fixed channel and fiber unpacking bug

ldmolacek committed:
Changes to QIE10/11DataFrame

Added copyContent

kpedro88 committed:
QIE10/11 dataframes in ECAL style (edm::DataFrame as member)

kpedro88 committed:
propagate QIE10DataFrame to all digitization classes (compiles!)

kpedro88 committed:
fix some configurations for QIE10

kpedro88 committed:
get correct number of samples from QIE10

kpedro88 committed:
revert testing scenario for QIE10 local reco

kpedro88 committed:
setup hcal digitizer for HF QIE8/10 shared case

kpedro88 committed:
setup hardcode conditions for HFQIE10 2016 test

kpedro88 committed:
add sub-eras to internal use list

kpedro88 committed:
update HF 2016 hardcode scenario

kpedro88 committed:
more informative LogError in HcalDigitizer

kpedro88 committed:
reduce verbosity of hcal unpacker

cmsbuild commented:
A new Pull Request was created by  @kpedro88 (Kevin Pedro) for CMSSW_8_0_X.

It involves the following packages:

CalibCalorimetry/HcalAlgos
CalibCalorimetry/HcalPlugins
CalibFormats/HcalObjects
CondFormats/HcalObjects
Configuration/EventContent
Configuration/StandardSequences
DataFormats/HcalDigi
EventFilter/HcalRawToDigi
RecoLocalCalo/HcalRecAlgos
RecoLocalCalo/HcalRecProducers
SimCalorimetry/CaloSimAlgos
SimCalorimetry/HcalSimAlgos
SimCalorimetry/HcalSimProducers
SimCalorimetry/HcalZeroSuppressionProducers
SimGeneral/MixingModule


@civanch, @diguida, @cvuosalo, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @slava77, @ggovi, @mmusich, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @apfeiffer1, @wmtan, @makortel, @GiacomoSguazzoni, @tocheng, @VinInn, @Martin-Grunewald, @rovere, @cerati, @argiro, @dgulhan this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on DataFormats/HcalDigi/interface/HcalDigiCollections.h:
```diff
@@ -35,16 +35,19 @@ typedef edm::SortedCollection<HcalTTPDigi> HcalTTPDigiCollection;
-  Digi operator[](size_type i) const { return Digi(edm::DataFrameContainer::operator[](i));}
+  typedef edm::DataFrameContainer::size_type size_type;
+  static const size_type MAXSAMPLES = 10;
+  HcalDataFrameContainer(int nsamples_per_digi=MAXSAMPLES, int isubdet=0) : edm::DataFrameContainer(nsamples_per_digi*Digi::WORDS_PER_SAMPLE+Digi::HEADER_WORDS+Digi::FLAG_WORDS, isubdet) { }
```
please put a linebreak.
This line is too long

slava77 commented on Configuration/EventContent/python/EventContent_cff.py:
```diff
@@ -359,6 +359,7 @@
                                                'keep HBHEDataFramesSorted_hcalDigis_*_*',
                                                'keep HFDataFramesSorted_hcalDigis_*_*',
                                                'keep HODataFramesSorted_hcalDigis_*_*',
+                                               'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_*_*',
```
Seeing how this shows up in the list of products, I think it would be much better to have "HF" in the class name or, alternatively, in the product instance.
Is there a chance we end up with QIE10 in more than just HF?

kpedro88 replied:
As far as I know, only HF will have QIE10. HB and HE will have QIE11 for their respective upgrades. The full name for, e.g., the output of simHcalUnsuppressedDigis ends up being:
```
QIE10DataFrameHcalDataFrameContainer_simHcalUnsuppressedDigis_HFQIE10DigiCollection_HLT
```

So the HF name does show up there.

slava77 replied:
On 2/9/16 1:37 PM, Kevin Pedro wrote:
> In Configuration/EventContent/python/EventContent_cff.py
> <https://github.com/cms-sw/cmssw/pull/13232#discussion_r52377863>:
> 
>> @@ -359,6 +359,7 @@
>>                                                 'keep HBHEDataFramesSorted_hcalDigis_*_*',
>>                                                 'keep HFDataFramesSorted_hcalDigis_*_*',
>>                                                 'keep HODataFramesSorted_hcalDigis_*_*',
>> +                                               'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_*_*',
> 
> As far as I know, only HF will have QIE10. HB and HE will have QIE11 for
> their respective upgrades. The full name for, e.g., the output of
> simHcalUnsuppressedDigis ends up being:
> 
> |QIE10DataFrameHcalDataFrameContainer_simHcalUnsuppressedDigis_HFQIE10DigiCollection_HLT
> |
> 
> So the HF name does show up there.

OK, that's good.

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13232/files#r52377863>.
>

slava77 commented on DataFormats/HcalDigi/interface/QIE11DataFrame.h:
```diff
@@ -3,13 +3,12 @@
 
  */
-class QIE11DataFrame : protected edm::DataFrame {
+class QIE11DataFrame {
```
do we have any samples with QIE11 (or QIE10) created?
I'm not sure the change in the inheritance is easily supported by the schema evolution.

kpedro88 replied:
These classes were created by @jmmans in 2015 and have not yet been used anywhere, until this PR. This change from inheritance to composition moves to the ECAL style of dataframe classes, which is a lot easier to handle due to some idiosyncrasies when trying to inherit from  edm::DataFrame.

slava77 commented on RecoLocalCalo/HcalRecAlgos/interface/HcalSimpleRecAlgo.h:
```diff
@@ -71,6 +72,7 @@ class HcalSimpleRecAlgo {
 
   HFRecHit reconstruct(const HFDataFrame& digi,  int first, int toadd, const HcalCoder& coder, const HcalCalibrations& calibs) const;
+  HFRecHit reconstructQIE10(const QIE10DataFrame& digi,  int first, int toadd, const HcalCoder& coder, const HcalCalibrations& calibs) const;
   HFRecHit reconstructHFUpgrade(const HcalUpgradeDataFrame& digi,  int first, int toadd, const HcalCoder& coder, const HcalCalibrations& calibs) const;
```
Is it me, or is there no difference between implementations of reconstruct, reconstructQIE10, and reconstructHFUpgrade methods?
Time to put a template instead?

kpedro88 replied:
Many interfaces in the HCAL digi and reco code could be simplified by templating, but I decided that this PR would already have so many changes, it was better not to confound things at the moment.

slava77 commented on RecoLocalCalo/HcalRecProducers/python/HFQIE10Reconstructor_cfi.py:
```diff
@@ -0,0 +1,15 @@
+import FWCore.ParameterSet.Config as cms
+
+hfQIE10Reco = cms.EDProducer("HcalSimpleReconstructor",
```
please keep the name of the module and the cfi file name the same.
hfQIE10Reco_cfi

slava77 commented on RecoLocalCalo/HcalRecProducers/src/HcalSimpleReconstructor.cc:
```diff
@@ -239,6 +245,46 @@ void HcalSimpleReconstructor::processUpgrade(edm::Event& e, const edm::EventSetu
+	  
+      // rof 27.03.09: drop ZS marked and passed digis:
+      if (dropZSmarkedPassed_)
+      if (frame.zsMarkAndPass()) continue;
```
please combine in one "if", add an indentation for the second if or put it in "{}" to improve readability.

slava77 commented on SimCalorimetry/HcalSimProducers/python/hcalUnsuppressedDigis_cfi.py:
```diff
@@ -50,6 +52,9 @@
 from Configuration.StandardSequences.Eras import eras
 if eras.fastSim.isChosen():
     hcalSimBlock.hitsProducer = cms.string('famosSimHits')
+if eras.run2_HF_2016.isChosen():
```
isChosen is not a good choice
@Dr15Jones please correct me if I'm wrong, and if not please suggest what's the best solution now.

kpedro88 replied:
I realize that isChosen is not ideal, but the problem is that the hcalSimBlock PSet is imported by multiple configs. If I use toModify, the changes will not get picked up. You can see that FastSim also has to do this...

makortel replied:
I leave the final word to @Dr15Jones, but actually `toModify()` applies the changes immediately (at least currently, and I've understood from Chris that this effect would be preserved in the future too).

Dr15Jones replied:
This should be changed to `toModify()`. Matti is right that `toModify()` is applied immediately. I would like to change that and if I did I would have to preserve the effect that the changes would also be picked up by anything importing that parameter.

kpedro88 replied:
@makortel @Dr15Jones - Thanks for the info about `toModify()`. I'll correct the usage in this PR (and also fix the FastSim one while I'm at it, attn @lveldere).

slava77 commented:
@kpedro88 
there are apparently only ~60 useful line changes in EventFilter/HcalRawToDigi/src/HcalUnpacker.cc ("788 additions, 729 deletions " that's here now is mostly from re-indentation).
Please go back to the minimal set of changes in the file.
The reasons to do it are
- to keep the modification history to an appropriate set of lines
- allow github web interface to display the changes so that comments can be added inline.

kpedro88 committed:
address first round of code review

cmsbuild commented:
Pull request #13232 was updated. @civanch, @diguida, @cvuosalo, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @slava77, @ggovi, @mmusich, @davidlange6 can you please check and sign again.

kpedro88 commented:
@slava77 - I addressed HcalUnpacker.cc and most of your other comments in the most recent commit. (The commits for HcalUnpacker were imported from someone else and I had not looked closely at them. I've also now commented out all the cout statements that were added.)

slava77 commented on EventFilter/HcalRawToDigi/src/HcalUnpacker.cc:
```diff
@@ -582,17 +582,62 @@ void HcalUnpacker::unpackUTCA(const FEDRawData& raw, const HcalElectronicsMap& e
+    //debug printouts
+    //int nwords=uhtr.getRawLengthBytes()/2;
+    //for (int iw=0; iw<nwords; iw++) 
+    //    printf("%04d %04x\n",iw,uhtr.getRawData16()[iw]);
```
generally, all commented out code should not be present.
If it's useful, it can be hidden by ifdefs, or converted to LogTrace or LogDebug.

slava77 commented:
@cmsbuild please test

@kpedro88 Please suggest a set of commands (or a config) to test QIE10 parts.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 2af08a03582e8f97f88e1109bd613f4f8092b820
 I found an error when building: 
 
 <pre>                 from /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/tuple:39,
                 from /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/functional:55,
                 from /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/memory:79,
                 from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw-patch/CMSSW_8_0_X_2016-02-09-1100/src/FWCore/MessageLogger/interface/MessageLogger.h:126,
                 from /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/EventFilter/CSCRawToDigi/src/CSCTMBScope.cc:8:
/afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/string:40:70: error: /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/char_traits.h: Bad address
 #include <bits/char_traits.h>  // NB: In turn includes stl_algobase.h
                                                                      ^
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/EventFilter/CSCRawToDigi/src/bitset_append.cc 
In file included from /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/string:52:0,
                 from /afs/cern.ch/cms/sw/ReleaseCandidates/volB/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/stdexcept:39,
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11106/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7c1d40335add64a7cd301dc092f48155c67f5c56
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11106/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11106/git-merge-result

slava77 commented:
@cmsbuild please test
looks like the last one was an issue with infrastructure

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 2af08a03582e8f97f88e1109bd613f4f8092b820
 I found an error when building: 
 
 <pre>>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisGenerator.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisL1Extra.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisL1Menu.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisL1Upgrade.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisRCT.cc 
c++: internal compiler error: Segmentation fault (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2016-02-09-1100/src/L1Trigger/L1TNtuples/src/L1AnalysisRecoCluster.cc 
gmake: *** [tmp/slc6_amd64_gcc493/src/L1Trigger/L1TNtuples/src/L1TriggerL1TNtuples/L1AnalysisL1Upgrade.o] Error 1
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11109/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7c1d40335add64a7cd301dc092f48155c67f5c56
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11109/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11109/git-merge-result

abdoulline commented:
May we have test(s) re-started?

mmusich commented:
please test
@abdoulline let's try again

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 2af08a03582e8f97f88e1109bd613f4f8092b820
 I found an error when building: 
 
 <pre>Leaving library rule at src/Validation/HcalHits/test
>> Building edm plugin tmp/slc6_amd64_gcc493/src/Validation/MuonRPCDigis/src/ValidationMuonRPCDigis/libValidationMuonRPCDigis.so
Leaving library rule at Validation/EcalRecHits
/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_X_2016-02-07-0000/lib/slc6_amd64_gcc493/libSimDataFormatsEncodedEventId.so: file not recognized: Bad address
>> Building shared library tmp/slc6_amd64_gcc493/src/Validation/MuonCSCDigis/src/ValidationMuonCSCDigis/libValidationMuonCSCDigis.so
collect2: error: ld returned 1 exit status
gmake: *** [tmp/slc6_amd64_gcc493/src/Validation/MuonRPCDigis/src/ValidationMuonRPCDigis/libValidationMuonRPCDigis.so] Error 1
Leaving library rule at Validation/MuonRPCDigis
>> Building edm plugin tmp/slc6_amd64_gcc493/src/Validation/Performance/test/SimDigiDumper/libSimDigiDumper.so
Copying tmp/slc6_amd64_gcc493/src/Validation/MuonCSCDigis/src/ValidationMuonCSCDigis/libValidationMuonCSCDigis.so to productstore area:
Leaving library rule at src/Validation/Performance/test
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11120/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7c1d40335add64a7cd301dc092f48155c67f5c56
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11120/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11120/git-merge-result

abdoulline commented:
Bad luck once again...
What might be the reason of this:
"libSimDataFormatsEncodedEventId.so: file not recognized: Bad address"   
- doesn't seem to be related to PR itself...

bsunanda commented:
@cmsbuild Please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 2af08a03582e8f97f88e1109bd613f4f8092b820
 I found an error when building: 
 
 <pre>>> Building edm plugin tmp/slc6_amd64_gcc493/src/Validation/EcalDigis/src/ValidationEcalDigis/libValidationEcalDigis.so
Leaving library rule at src/L1Trigger/CSCTrackFinder/plugins
>> Building edm plugin tmp/slc6_amd64_gcc493/src/Validation/HcalHits/src/ValidationHcalHits/libValidationHcalHits.so
/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_X_2016-02-07-0000/lib/slc6_amd64_gcc493/libFWCoreParameterSet.so: file not recognized: Bad address
Leaving library rule at src/L1Trigger/TextToDigi/plugins
collect2: error: ld returned 1 exit status
>> Building big object file tmp/slc6_amd64_gcc493/src/Validation/HcalHits/src/ValidationHcalHits/bigobj/ValidationHcalHits.obj
gmake: *** [tmp/slc6_amd64_gcc493/src/Validation/HcalHits/src/ValidationHcalHits/libValidationHcalHits.so] Error 1
Leaving library rule at src/L1Trigger/CSCTrackFinder/test
>> Building edm plugin tmp/slc6_amd64_gcc493/src/Validation/HcalHits/test/testValidationHcalHits/libtestValidationHcalHits.so
Leaving library rule at L1Trigger/L1TMuon
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11129/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7c1d40335add64a7cd301dc092f48155c67f5c56
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11129/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11129/git-merge-result

kpedro88 commented:
@slava77, @abdoulline - I've been getting these sporadic scram/gcc errors when compiling privately as well: bad address, segmentation faults in gcc, etc. Usually if I try a few times it works. I have no idea why this is happening; it doesn't seem to be related to any of the actual changes I've made in the PR.

slava77 commented:
@kpedro88 
We recently had a large fraction of issues with IBs from afs distribution.
for local compilation, I think you can move to cvmfs distribution instead of afs
/cvmfs/cms-ib.cern.ch/2016-{pick the latest relevant}/cmsset_default.sh  should put the IBs in the path

kpedro88 committed:
put HcalUnpacker cout statements behind ifdefs

kpedro88 committed:
use toModify instead of isChosen

cmsbuild commented:
Pull request #13232 was updated. @civanch, @diguida, @cvuosalo, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @slava77, @ggovi, @mmusich, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test
... expect only partial success due to the PR testing infrastructure issues

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented:
@slava77 - I didn't realize the IBs were available on cvmfs. I will keep that in mind next time I try to compile (I finally got it to compile on afs last night and would rather stick with the working version for the moment).

Testing the QIE10 functionality is actually a bit complicated. The Run2_2016 scenario relies on some database values that currently can only be set with hardcode conditions, and also requires #13139 and a forthcoming PR on top of that. Here is a simpler test you can do; this just generates pions in HF and then forces HcalDigitizer to use QIE10 for all HF cells.
```
 cmsDriver.py SinglePiE50HCAL_pythia8_cfi  --conditions auto:run2_mc -n 10 --era Run2_25ns --eventcontent FEVTDEBUG -s GEN,SIM --datatier GEN-SIM --beamspot Realistic50ns13TeVCollision --fileout file:step1.root --customise_commands "process.generator.PGunParameters.MinEta = cms.double(3.0)" > step1_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1
 
 cmsDriver.py step2  --conditions auto:run2_mc -s DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval25ns --datatier GEN-SIM-DIGI-RAW-HLTDEBUG -n 10 --era Run2_25ns --eventcontent FEVTDEBUGHLT --filein file:step1.root  --fileout file:step2.root --customise_commands "process.mix.digitizers.hcal.HFQIE8 = cms.bool(False)\nprocess.mix.digitizers.hcal.HFQIE10 = cms.bool(True)" > step2_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1
 
 cmsDriver.py step3  --runUnscheduled  --conditions auto:run2_mc -s RAW2DIGI,L1Reco,RECO,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@miniAODDQM --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO -n 10 --era Run2_25ns --eventcontent RECOSIM,MINIAODSIM,DQM --filein file:step2.root  --fileout file:step3.root > step3_SinglePiE50_UP15+SinglePiE50_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log  2>&1
```
You'll be able to see that the QIE10 collections are filled and the HFDataFrame collections are empty in step2.root. The placeholder reconstruction algorithm for QIE10 isn't enabled right now (since it won't be used in 2016).

If you want to try this more easily/just see the results, you can load from my lxplus area:
```
/afs/cern.ch/work/p/pedrok/public/hf/rebase2/CMSSW_8_0_X_2016-02-09-1100/src
```
The results are in the `test` directory.

cmsbuild commented:
-1 
Tested at: 8a3468dbc55477b10e73d911796a617044ae6860
 I found an error when building: 
 
 <pre>>> Building edm plugin tmp/slc6_amd64_gcc493/src/RecoLocalCalo/HcalRecAlgos/plugins/RecoLocalCaloHcalRecAlgosPlugin/libRecoLocalCaloHcalRecAlgosPlugin.so
--- Registered EDM Plugin: HLTriggerHLTanalyzers
>> Leaving Package DQM/SiPixelMonitorDigi
>> Package DQM/SiPixelMonitorDigi built
/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_X_2016-02-07-0000/lib/slc6_amd64_gcc493/libDataFormatsEcalDetId.so: error adding symbols: Bad address
collect2: error: ld returned 1 exit status
@@@@ Running edmWriteConfigs for DQMHcalMonitorClient
gmake: *** [tmp/slc6_amd64_gcc493/src/RecoLocalCalo/HcalRecProducers/src/RecoLocalCaloHcalRecProducers/libRecoLocalCaloHcalRecProducers.so] Error 1
Leaving library rule at src/CondTools/Hcal/plugins
>> Leaving Package DQM/SiPixelMonitorTrack
>> Package DQM/SiPixelMonitorTrack built
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11136/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7c1d40335add64a7cd301dc092f48155c67f5c56
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11136/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11136/git-merge-result

kpedro88 commented:
@slava77 @mmusich @bsunanda  - can you relaunch the tests if the IB problems are sorted out now?

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 8a3468dbc55477b10e73d911796a617044ae6860
 I found an error when building: 
 
 <pre>>> Package RecoBTag/Skimming built
@@@@ Running edmWriteConfigs for testTrackingToolsTrackAssociator
--- Registered EDM Plugin: RecoLocalMuonCSCRecHitD
Copying tmp/slc6_amd64_gcc493/src/CondCore/Utilities/bin/cmscond_from_lumiid/cmscond_from_lumiid to productstore area:
/afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_X_2016-02-10-2300/external/slc6_amd64_gcc493/lib/liblcg_CoralKernel.so: file not recognized: Bad address
collect2: error: ld returned 1 exit status
gmake: *** [tmp/slc6_amd64_gcc493/src/DQMServices/Diagnostic/src/DQMServicesDiagnostic/libDQMServicesDiagnostic.so] Error 1
--- Registered EDM Plugin: DQMServicesFileIOPlugins
@@@@ Running edmWriteConfigs for testCaloMatchingExample
--- Registered EDM Plugin: testTrackingToolsTrackAssociator
@@@@ Running edmWriteConfigs for TrackingToolsTrackAssociatorPlugins
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11148/summary.html

slava77 commented on RecoLocalCalo/HcalRecProducers/python/hfQIE10Reco_cfi.py:
```diff
@@ -0,0 +1,15 @@
+
+hfQIE10Reco = cms.EDProducer("HcalSimpleReconstructor",
+    correctionPhaseNS = cms.double(0.0),
+    digiLabel = cms.InputTag("simHcalUnsuppressedDigis","HFQIE10DigiCollection"),
```
From the ORP on Tuesday I understood that there will be a full proper digi2raw -> raw2digi chain and the default for QIE10 inputs will be just the hcalDigis.
This setup looks like a temporary bypass for a case of a missing packer (and no expected actual detector data).
Please clarify the situation.

kpedro88 replied:
This cfi and the corresponding algorithm in HcalSimpleReconstructor are just placeholders to demonstrate the use of the QIE10DigiCollection. This isn't in the RECO workflow for any scenario and never will be. Eventually it will be supplanted by the real algorithms, once they are developed.

slava77 replied:
On 2/11/16 1:10 PM, Kevin Pedro wrote:
> In RecoLocalCalo/HcalRecProducers/python/hfQIE10Reco_cfi.py
> <https://github.com/cms-sw/cmssw/pull/13232#discussion_r52667143>:
> 
>> @@ -0,0 +1,15 @@
>> +import FWCore.ParameterSet.Config as cms
>> +
>> +hfQIE10Reco = cms.EDProducer("HcalSimpleReconstructor",
>> +    correctionPhaseNS = cms.double(0.0),
>> +    digiLabel = cms.InputTag("simHcalUnsuppressedDigis","HFQIE10DigiCollection"),
> 
> This cfi and the corresponding algorithm in HcalSimpleReconstructor are
> just placeholders to demonstrate the use of the QIE10DigiCollection.
> This isn't in the RECO workflow for any scenario and never will be.
> Eventually it will be supplanted by the real algorithms, once they are
> developed.

I had another question in the comment: do we have an operational
digi2raw (packer) for simulation?

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13232/files#r52667143>.
>

kpedro88 replied:
I had assumed that @jmmans and co. had handled this already, but now that I look, it appears that HcalPacker has not yet been modified to handle QIE10 digis. We should take care of this, but I hope it can come in a subsequent PR.

awhitbeck commented:
The unpacker is definitely on our radar, but it is delayed.  I hope we'll have a first version in a week or so.

slava77 commented:
On 2/11/16 1:26 PM, Andrew Whitbeck wrote:
> The unpacker is definitely on our radar, but it is delayed. I hope we'll
> have a first version in a week or so.
> 

Good to know.
We can move on with this PR otherwise.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13232#issuecomment-183065293>.
>

slava77 commented:
Regarding detector data: when is the QIE10 unpacking code going to have to become operational?
I suppose a channel map in GT will enable unpacking of relevant channels.
When are the changes in this PR in the unpacker become a requirement to deliver the needed hcalDigis
from the data taking perspective?
Is it MWGR2 ?

abdoulline commented:
Unpacker is not the issue for QIE10,  Packer (MC) is, as Kevin has 
mentioned.
Once "extended" emap is available (being prepared, goal is to have it by 
MWGR2) QIE10 Digis will be read out. At least privately, if central GT will 
not include updated emap by then.


On Thu, 11 Feb 2016, Slava Krutelyov wrote:

> 
> Regarding detector data: when is the QIE10 unpacking code going to have to become operational?
> I suppose a channel map in GT will enable unpacking of relevant channels.
> When are the changes in this PR in the unpacker become a requirement to deliver the needed hcalDigis
> from the data taking perspective?
> Is it MWGR2 ?
> 
> —
> Reply to this email directly or view it on GitHub.[AEx02rAaIJ923cZ6EA5XtKpC2HEixzJ4ks5pjPfIgaJpZM4HWi_Q.gif]
> 
> 
>

kpedro88 commented:
Yes, there are a lot of things that need to be put in the database and GT for the 2016 scenario to work: not just the emap, but also all of the QIETypes, calibration constants, etc. for the "new" cells in HF. The new geometry/topology for HF will also have to be specified (currently in the works).

abdoulline commented:
Agree, I meant "bare-bones": emap for unpacking first.

On Thu, 11 Feb 2016, Kevin Pedro wrote:

> 
> Yes, there are a lot of things that need to be put in the database and GT for the 2016 scenario to work: not just
> the emap, but also all of the QIETypes, calibration constants, etc. for the "new" cells in HF. The new
> geometry/topology for HF will also have to be specified (currently in the works).
> 
> —
> Reply to this email directly or view it on GitHub.[AEx02iqMO0xXwrF9Wp5YYIsOE2kKmgyaks5pjPqigaJpZM4HWi_Q.gif]
> 
> 
>

slava77 commented:
@cmsbuild  please test
what if ...

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented:
@slava77 - I'm still getting random segmentation faults in gcc when compiling from the cvmfs IB instead of the afs IB.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/13232	8a3468d
- changes in the code are somewhat in line with the description: unpacking of QIE11 is also added
- jenkins tests were not particularly successful, so, I used local tests in CMSSW_8_0_X_2016-02-11-1100: tests (compilation regular and clang, runtests, addOn tests, and matrix workflows) pass and comparisons with baseline show no differences in run1/run2 (2015) matrix workflows; the 2017 workflows are broken in the IB and were not checked
    - step2 now creates QIE10DataFrameHcalDataFrameContainer simHcal\*Digis_HFQIE10DigiCollection, which is empty (as expected for the supplied channel map)
    - tests suggested in https://github.com/cms-sw/cmssw/pull/13232#issuecomment-182506645 pass and create digi collections, the simHcalUnsuppressedDigis_HFQIE10DigiCollection has a size of 38016.
    - step3 still runs on the baseline step2 outputs, to check that the DataFormats changes didn't generate unexpected issues.

slava77 commented:
@cmsbuild please test
based on the jenkins console https://cmssdt.cern.ch/jenkins/job/ib-any-integration/11152/console, it looks like the compilation got stuck (it can't really take 10 hours to compile in a normal environment)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 8a3468dbc55477b10e73d911796a617044ae6860
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11176/summary.html

kpedro88 commented:
Looks like a spurious failure in the unit test (has happened with other PRs recently as well).

kpedro88 commented:
@slava77 @bsunanda @mmusich - can we restart the tests? This has already been delayed substantially by the IB problems and it's important to get it into 8_0_0.

@ggovi @civanch - Can you also look over this PR ASAP?

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 8a3468dbc55477b10e73d911796a617044ae6860
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11226/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11226/summary.html

kpedro88 committed:
use fillDescriptions in HcalHardcodeCalibrations

cmsbuild commented:
Pull request #13232 was updated. @civanch, @diguida, @cvuosalo, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @slava77, @ggovi, @mmusich, @davidlange6 can you please check and sign again.

kpedro88 commented:
@slava77 @mmusich - The listed failure in testRecoMETMETProducers appears, again, to be spurious. (Why do we keep tests that cause so many false positives?)

However, I noticed that there was a real failure in some of the add-on tests, because I added a parameter to HcalHardcodeCalibrations. Following previous recommendations, I added a `fillDescriptions()` function so the parameters will be handled correctly in the HLT configs. The add-on tests should pass now, please restart the tests when you can.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on CalibCalorimetry/HcalPlugins/python/Hcal_Conditions_forGlobalTag_cff.py:
```diff
@@ -5,20 +5,22 @@
 )
 
+from Configuration.StandardSequences.Eras import eras
+if eras.run2_HF_2016.isChosen():
```
I thought the goal was to use toModify.
The last commit undid it 3fa6f8d2a

kpedro88 replied:
@slava77 - Thanks for the catch, fixed.

kpedro88 committed:
restore toModify

cmsbuild commented:
Pull request #13232 was updated. @civanch, @diguida, @cvuosalo, @cerminar, @cmsbuild, @franzoni, @mdhildreth, @slava77, @ggovi, @mmusich, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 1b8d4b93d1543eba06415b03dc2dfbf33a495cd8
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11239/summary.html

kpedro88 commented:
Unit test failure was spurious again. Matrix test failures are both DAS errors.

kpedro88 commented:
I decided to investigate testRecoMETMETProducers because its persistent failures have become irritating. I don't think it's actually possible for it to work as designed. It looks for the un-cleaned `particleFlowRecHitHF` collection, but this is not kept in any standard EventContent.
[`process.particleFlowClusterHF`](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoMET/METProducers/test/recoMET_pfClusterMet_cfg.py#L37)
[`recHitsSource = cms.InputTag("particleFlowRecHitHF")`](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHF_cfi.py#L106)
[`--eventcontent FEVTDEBUGHLT`](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoMET/METProducers/test/runtests.sh#L10)
[`'keep recoPFRecHits_particleFlowRecHitHF_Cleaned_*'`](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoParticleFlow/Configuration/python/RecoParticleFlow_EventContent_cff.py#L22)
This should be fixed somehow rather than causing problems with every relevant PR...

slava77 commented:
@cmsbuild please test
unfortunately, the DAS errors are blocking the comparisons
the unit test errors are non-blocking

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
On 2/15/16 9:54 PM, Kevin Pedro wrote:
> I decided to investigate testRecoMETMETProducers because its persistent
> failures have become irritating. I don't think it's actually possible
> for it to work as designed. It looks for the un-cleaned
> |particleFlowRecHitHF| collection, but this is not kept in any standard
> EventContent.
> |process.particleFlowClusterHF|
> <https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoMET/METProducers/test/recoMET_pfClusterMet_cfg.py#L37>
> |recHitsSource = cms.InputTag("particleFlowRecHitHF")|
> <https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHF_cfi.py#L106>
> |--eventcontent FEVTDEBUGHLT|
> <https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoMET/METProducers/test/runtests.sh#L10>
> |'keep recoPFRecHits_particleFlowRecHitHF_Cleaned_*'|
> <https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/RecoParticleFlow/Configuration/python/RecoParticleFlow_EventContent_cff.py#L22>
> This should be fixed somehow rather than causing problems with every
> relevant PR...


Hi Kevin,

Thank you for taking a look.
Occasionally, the same PR was failing due to the tile itself to be created.
The issue with event content is clearly predictable and I will provide a
fix to this.


Slava

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13232#issuecomment-184532914>.
>

cmsbuild commented:
-1 
 
Tested at: 1b8d4b93d1543eba06415b03dc2dfbf33a495cd8
I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11246/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13232/11246/summary.html

kpedro88 commented:
The comparison looks good to me - no changes to any existing workflows.

While I'm complaining about pointless tests:
* Why do we run the DQM TimerService when it always has a large percentage of expected failures?
* Is it possible to tweak the BinToBin calculation so it doesn't sometimes decide that a comparison of 0.0 to 0.0 is a failure?
* @lveldere @ssekmen - it would be useful if the FastSim problem with HcalHitsV (present for ~3 months now) could be diagnosed and fixed...

davidlange6 commented:
@kpedro88 - I agree - looks good in the comparison

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/13232 1b8d4b9
- changes since the last signoff affect only the HcalHardcodeCalibrations https://github.com/kpedro88/cmssw/compare/8a3468d...1b8d4b93 
- jenkins tests pass and comparisons with baseline show no differences

kpedro88 commented:
@davidlange6 - the QIE10s are indeed in the MWGR.

@mmusich @civanch @ggovi - please review and sign ASAP.

ggovi commented on CondFormats/HcalObjects/interface/HcalQIEType.h:
```diff
@@ -12,6 +12,7 @@ contains the QIE Typese + corresponding DetId
 class HcalQIEType
 {
  public:
+  enum {QIE8=0, QIE10=1, QIE11=2};
```
Do you really need to have this enum in the condFormat? I guess this won't be used when the HcalQIEType payloads will be read from the DB... Maybe you could place this enum in a namespace and leave the CondFormat unchanged.

kpedro88 replied:
I wasn't sure where to put it. Currently I use the enum in HcalDigitizer and HcalDbHardcode. Conceptually, the enum is used to make the HcalQIEType values human readable/understandable, but if there is a better/more standard location to define it, let me know.

ggovi replied:
One of the most important 'standard' for the persistency is the fact that the CondFormat classes are frozen = no change possible. This case is actually borderline, since it does not involve persistent data. However, and especially if in the future you won't be using the enum anymore, you might consider to simply put it in a namespace... What about HcalDbHardcode ?

kpedro88 replied:
Indeed, I thought it was acceptable to add the enum to the CondFormat since it is not persistent.

If anything, we will use the enum more heavily in the future, as the HCAL upgrades progress and more new QIEs are installed. I do not think HcalDbHardcode is the right place to define it, since the enum will be used to interpret *both* database and hardcode conditions.

kpedro88 replied:
It should also be noted that QIEType was only added to 80X ~2 months ago in #12473, so it really hasn't yet been used anywhere centrally and minor modifications at this stage won't cause any persistency problems.

franzoni commented on CalibCalorimetry/HcalAlgos/src/HcalDbHardcode.cc:
```diff
@@ -564,6 +564,28 @@ HcalQIECoder HcalDbHardcode::makeQIECoder (HcalGenericDetId fId) {
+    if(fId.genericSubdet() == HcalGenericDetId::HcalGenForward && fId.isHcalDetId()) {
+        HcalDetId hid(fId);
+        if(hid.depth()>=3) qieType = HcalQIEType::QIE10;
+    }
```
Hello @kpedro88 

we understand that testHFQIE10 serves the purpose of switching at python level to the 2016 scenario (if set to true), where only a few channels need be handled with the QIE10.
If set to false, the HCAL is seen as >=2017, after the full transition to the new readout.

Why is depth >=3 used as discriminant to set QIE10 ?
According to https://indico.cern.ch/event/492087/contribution/4/attachments/1225068/1792949/Phase2-Talk30.pdf  slide 4, only HB can have 3 levels of segmentation, not HF.

Furthermore, and more critically:
who is data <=2015 going to be handled ? We understand that this PR is not the final word - given than no DB content is actually used, no history influences how the HcalQIEType  are set up; it would teherefor appear that when running on data from the past, the HcalQIEType will be incorrectly set to QIE10.
Can you explain, pls ?

@ggovi @mmusich

kpedro88 replied:
@franzoni - In the 2016 HF, the new QIE10 readouts/cells are marked as depths 3/4 (corresponding to the existing HF depths 1/2). This is allowed and supported by the new topology code. Slide 4 in the talk you reference is "till 2015", i.e. it does not address the 2016+ detector. It is important to note that depths 3/4 will never survive to the RECO stage, so HF at the RecHit level will still only have depths 1/2. In 2016, the QIE10 cells will just be ignored for RECO; in 2017, we will have full dual anode readout, and then depths 1 and 3 (2 and 4) will be combined intelligently to make a single RecHit.

HcalDbHardcode is only used for testing purposes. It will not be used for the "old" scenarios. You can further see that the default is set to QIE8:
```
int qieType = HcalQIEType::QIE8; //default
```
and the `testHFQIE10` flag is only enabled with the 2016 era. There should not be any issues maintaining the old scenarios.

davidlange6 commented:
@kpedro88 - maybe its wise to get a PR with the Data format changes and enough of the unpacker changes so that MWGR#2 data will run without crashing? The cond changes appear to need more discussion.

kpedro88 commented:
@davidlange6 - I just talked with @awhitbeck about the MWGR in more detail. We are also missing some necessary conditions for unpacking the QIE10 data (emap in DB and list of new FEDs). Given that, we think it might be a more feasible plan just to store only the RAW data and then unpack it privately, offline, after the MWGR. Then the review of this PR can proceed.

(I am not thrilled by the idea of trying a not-very-straightforward split into multiple PRs. In all likelihood, they would only end up taking even longer to be approved. This PR has already been open for > week and as usual, the review did not happen as promptly as it could have.)

If there is a standard place to define an enum which describes database values, I am happy to move the definition to satisfy @ggovi's comment. It needs to be sufficiently low-level that it can be included wherever necessary without breaking any dependency rules. My expectation was that without this enum, I would likely get a review comment discouraging the use of magic numbers.

I think that @franzoni's comment was just a misunderstanding about when the hardcode conditions will actually be used. I hope he will clarify if my explanation was sufficient.

davidlange6 commented:
I think I can approve the revo part for 800,the rest not. people need doom time to transition to a major release cycle... Even if a bug fix comes later
________________________________________
From: Kevin Pedro [notifications@github.com]
Sent: 17 February 2016 20:40
To: cms-sw/cmssw
Cc: David Lange
Subject: Re: [cmssw] HF QIE10 for 2016 (#13232)

@davidlange6<https://github.com/davidlange6> - I just talked with @awhitbeck<https://github.com/awhitbeck> about the MWGR in more detail. We are also missing some necessary conditions for unpacking the QIE10 data (emap in DB and list of new FEDs). Given that, we think it might be a more feasible plan just to store only the RAW data and then unpack it privately, offline, after the MWGR. Then the review of this PR can proceed.

(I am not thrilled by the idea of trying a not-very-straightforward split into multiple PRs. In all likelihood, they would only end up taking even longer to be approved. This PR has already been open for > week and as usual, the review did not happen as promptly as it could have.)

If there is a standard place to define an enum which describes database values, I am happy to move the definition to satisfy @ggovi<https://github.com/ggovi>'s comment. It needs to be sufficiently low-level that it can be included wherever necessary without breaking any dependency rules. My expectation was that without this enum, I would likely get a review comment discouraging the use of magic numbers.

I think that @franzoni<https://github.com/franzoni>'s comment was just a misunderstanding about when the hardcode conditions will actually be used. I hope he will clarify if my explanation was sufficient.

�
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/13232#issuecomment-185370603>.

kpedro88 commented:
@davidlange6 - I think your last comment was mangled by autocorrect, could you clarify?

davidlange6 commented:
Indeed.....

I can approve the reco part "now" so we can survive the Mwgr... We need 800 now so that tier0 and hot can transition., we can make bug fixes later

kpedro88 commented:
@davidlange6 - well, according to my conversation with @awhitbeck, we can "survive" the MWGR by just storing the RAW data (no code changes necessary). It would be nice to be able to unpack the data online, but because of the other missing things (emap + FED list, on top of the dataformat/unpacker changes in this PR) it's probably not feasible anyway.

If I try to split off the dataformat/unpacker changes from this PR (part A), the remainder of the PR (part B) cannot even be *submitted* until part A is merged, because of dependence on new object methods, etc. Then 80X will already be closed for new features, so part B will have to be submitted to 81X and then backported. I cannot spend forever just tweaking this PR; we need to integrate it so we can deal with the next issues for HCAL, of which there are many (and the schedule seems to accelerate by several weeks every time I talk to anyone). Again, we are in this situation because of IB/testing issues that are totally outside my control. This has happened before and somehow it is always the developers who are penalized for release management problems.

davidlange6 commented:
Right, but by I can spprove no,w I mean now ....don't worry about integrating into 81x just yet. 
________________________________________
From: Kevin Pedro [notifications@github.com]
Sent: 17 February 2016 21:17
To: cms-sw/cmssw
Cc: David Lange
Subject: Re: [cmssw] HF QIE10 for 2016 (#13232)

@davidlange6<https://github.com/davidlange6> - well, according to my conversation with @awhitbeck<https://github.com/awhitbeck>, we can "survive" the MWGR by just storing the RAW data (no code changes necessary). It would be nice to be able to unpack the data online, but because of the other missing things (emap + FED list, on top of the dataformat/unpacker changes in this PR) it's probably not feasible anyway.

If I try to split off the dataformat/unpacker changes from this PR (part A), the remainder of the PR (part B) cannot even be submitted until part A is merged, because of dependence on new object methods, etc. Then 80X will already be closed for new features, so part B will have to be submitted to 81X and then backported. I cannot spend forever just tweaking this PR; we need to integrate it so we can deal with the next issues for HCAL, of which there are many (and the schedule seems to accelerate by several weeks every time I talk to anyone). Again, we are in this situation because of IB/testing issues that are totally outside my control. This has happened before and somehow it is always the developers who are penalized for release management problems.

�
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/13232#issuecomment-185384760>.

davidlange6 commented:
No,w" now,
Sapprove= approve

abdoulline commented:
Yes, we agreed with HCAL OPS that for MWGR #2 we'll have only QIE8 channels in emap (and replacing all HBHE VME channels with uTCA ones), then after it - will update emap with QIE10 channels. Extension of FEDs for unpacking - to be discussed. It might need to explicitly add them to the unpacker - to be clarified tomorrow.

kpedro88 commented:
superseded by #13331 and #13334.

franzoni commented:
thanks for the replies and explications.

concerning:

> This PR has already been open for > week and as usual, the review did not happen as promptly as it 
> could have.

Reviewing a PR of this size being only partly aware of the contextual details
and having to back-engineer from the code the details of the plan you're executing,
takes an amount of time and effort much much larger than needed.

I made this point already a few days back, when we asked a report at AlCa (by Sunanda). Reminder to the HCAL DPG team: if you want to go "fast" when deploying changes with this level of implications,
you can help the process if you somehow make others aware of what you're trying to do and github is left for addressing the details.
